File: app/android/app/src/main/kotlin/com/sriannamalai/app/VaultChannel.kt
package com.sriannamalai.app

import android.app.Activity
import android.content.Intent
import android.net.Uri
import androidx.documentfile.provider.DocumentFile
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * The Android half of the `mdviewer/vault` platform channel — folder access
 * via Storage Access Framework (`ACTION_OPEN_DOCUMENT_TREE` +
 * `takePersistableUriPermission`), listing/reading through a
 * [DocumentFile] tree. See
 * `app/lib/src/vault/platform_vault_provider.dart` for the Dart-side
 * contract this implements: `pickFolder` / `restore` / `list` / `read`.
 *
 * Unlike the iOS side, SAF grants don't need an explicit "start/stop scope"
 * — a persisted `content://` tree Uri is usable directly from
 * [android.content.ContentResolver] any time the permission still holds, so
 * every method here just re-derives a [DocumentFile] from the `id` (the
 * tree Uri as a string) it's given rather than caching "the active" one.
 *
 * [MainActivity] owns the actual `ACTION_OPEN_DOCUMENT_TREE` activity
 * result (Android's picker is a `startActivityForResult` round trip, not a
 * synchronous call) and forwards it to [onActivityResult].
 *
 * This channel's actual folder-picker/SAF behavior only runs on a real
 * device or emulator — it is device-E2E-verified, not covered by
 * `flutter test` (see the task-2 report).
 */
class VaultChannel(private val activity: Activity, messenger: BinaryMessenger) : MethodChannel.MethodCallHandler {
  companion object {
    private const val CHANNEL_NAME = "mdviewer/vault"
    private const val REQUEST_CODE_PICK_FOLDER = 42_017

    /** Directories deeper than this below the picked root aren't descended into — mirrors MDViewer.Desktop's `read_dir_tree` depth bound. */
    private const val MAX_DEPTH = 6
  }

  private val channel = MethodChannel(messenger, CHANNEL_NAME)
  private var pendingPickResult: MethodChannel.Result? = null

  init {
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "pickFolder" -> pickFolder(result)

      "restore" -> {
        val id = call.argument<String>("id")
        if (id == null) {
          result.error("bad_args", "restore requires {id}", null)
        } else {
          result.success(restore(id))
        }
      }

      "list" -> {
        val id = call.argument<String>("id")
        if (id == null) {
          result.error("bad_args", "list requires {id}", null)
        } else {
          result.success(list(id))
        }
      }

      "read" -> {
        val id = call.argument<String>("id")
        val relPath = call.argument<String>("relPath")
        if (id == null || relPath == null) {
          result.error("bad_args", "read requires {id, relPath}", null)
        } else {
          val bytes = read(id, relPath)
          if (bytes == null) result.error("not_found", "could not read \"$relPath\"", null) else result.success(bytes)
        }
      }

      else -> result.notImplemented()
    }
  }

  // region pickFolder

  private fun pickFolder(result: MethodChannel.Result) {
    if (pendingPickResult != null) {
      result.error("busy", "a pickFolder() call is already in progress", null)
      return
    }
    pendingPickResult = result
    val intent =
      Intent(Intent.ACTION_OPEN_DOCUMENT_TREE).addFlags(
        Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION
      )
    activity.startActivityForResult(intent, REQUEST_CODE_PICK_FOLDER)
  }

  /** Forwarded from [MainActivity.onActivityResult]. */
  fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
    if (requestCode != REQUEST_CODE_PICK_FOLDER) return
    val result = pendingPickResult ?: return
    pendingPickResult = null

    val uri = if (resultCode == Activity.RESULT_OK) data?.data else null
    if (uri == null) {
      result.success(null) // user cancelled
      return
    }

    activity.contentResolver.takePersistableUriPermission(
      uri,
      Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION,
    )

    val doc = DocumentFile.fromTreeUri(activity, uri)
    if (doc == null || !doc.isDirectory) {
      result.error("invalid_folder", "picked Uri is not a folder", null)
      return
    }
    val displayName = doc.name ?: uri.lastPathSegment ?: "Folder"
    result.success(mapOf("id" to uri.toString(), "displayName" to displayName))
  }

  // endregion

  private fun restore(id: String): Boolean {
    val uri = parseUri(id) ?: return false
    val hasPermission =
      activity.contentResolver.persistedUriPermissions.any { it.uri == uri && it.isReadPermission }
    if (!hasPermission) return false // revoked: the Dart side drops the persisted grant and re-prompts
    val doc = DocumentFile.fromTreeUri(activity, uri) ?: return false
    return doc.isDirectory && doc.exists()
  }

  private fun list(id: String): List<String> {
    val uri = parseUri(id) ?: return emptyList()
    val root = DocumentFile.fromTreeUri(activity, uri) ?: return emptyList()
    val out = mutableListOf<String>()
    walk(root, "", MAX_DEPTH, out)
    return out
  }

  private fun walk(dir: DocumentFile, relPath: String, depth: Int, out: MutableList<String>) {
    if (depth <= 0) return
    val children = runCatching { dir.listFiles() }.getOrNull() ?: return
    for (child in children) {
      val name = child.name ?: continue
      if (name.startsWith(".")) continue // dotfiles/dot-dirs skipped, desktop rule
      val childRelPath = if (relPath.isEmpty()) name else "$relPath/$name"
      if (child.isDirectory) {
        walk(child, childRelPath, depth - 1, out)
      } else if (name.lowercase().endsWith(".md") || name.lowercase().endsWith(".markdown")) {
        out.add(childRelPath)
      }
    }
  }

  private fun read(id: String, relPath: String): ByteArray? {
    val uri = parseUri(id) ?: return null
    var node = DocumentFile.fromTreeUri(activity, uri) ?: return null
    val segments = relPath.split("/").filter { it.isNotEmpty() }
    if (segments.isEmpty()) return null
    for (segment in segments) {
      node = node.findFile(segment) ?: return null
    }
    if (node.isDirectory) return null
    return runCatching { activity.contentResolver.openInputStream(node.uri)?.use { it.readBytes() } }.getOrNull()
  }

  private fun parseUri(id: String): Uri? = runCatching { Uri.parse(id) }.getOrNull()
}

File: app/ios/Runner/VaultChannel.swift
import Flutter
import UIKit
import UniformTypeIdentifiers

/// The iOS half of the `mdviewer/vault` platform channel — folder access via
/// `UIDocumentPickerViewController` (folder mode, "open" not "import" so the
/// picked folder is never copied into the app sandbox) + security-scoped
/// bookmarks. See `app/lib/src/vault/platform_vault_provider.dart` for the
/// Dart-side contract this implements: `pickFolder` / `restore` / `list` /
/// `read`.
///
/// Only one folder is "active" (its security scope started) at a time,
/// matching the Dart-side `PlatformVaultProvider`'s one-grant-at-a-time
/// model (v1 non-goal: multiple vaults). `pickFolder`/`restore` stop any
/// previously-active scope before starting the new one so scopes never leak
/// across a folder switch.
///
/// This channel's actual folder-picker/bookmark behavior only runs on a
/// real device or simulator — it is device-E2E-verified, not covered by
/// `flutter test` (see the task-2 report).
final class VaultChannel: NSObject {
  static let channelName = "mdviewer/vault"

  /// Directories more than this many levels below the picked root are not
  /// descended into — mirrors MDViewer.Desktop's `read_dir_tree` depth
  /// bound.
  private static let maxDepth = 6

  private let channel: FlutterMethodChannel
  private var activeURL: URL?
  private var pendingPickCompletion: ((URL?) -> Void)?

  init(messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(name: Self.channelName, binaryMessenger: messenger)
    super.init()
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call, result: result)
    }
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "pickFolder":
      pickFolder(result: result)

    case "restore":
      guard let id = stringArg(call, "id") else {
        result(badArgs("restore requires {id}"))
        return
      }
      result(restore(id: id))

    case "list":
      guard let id = stringArg(call, "id") else {
        result(badArgs("list requires {id}"))
        return
      }
      result(list(id: id))

    case "read":
      guard let id = stringArg(call, "id"), let relPath = stringArg(call, "relPath") else {
        result(badArgs("read requires {id, relPath}"))
        return
      }
      result(read(id: id, relPath: relPath))

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func stringArg(_ call: FlutterMethodCall, _ key: String) -> String? {
    (call.arguments as? [String: Any])?[key] as? String
  }

  private func badArgs(_ message: String) -> FlutterError {
    FlutterError(code: "bad_args", message: message, details: nil)
  }

  // MARK: - pickFolder

  private func pickFolder(result: @escaping FlutterResult) {
    guard pendingPickCompletion == nil else {
      // Mirrors the Android side: without this, a second concurrent call
      // would overwrite pendingPickCompletion, orphaning the first call's
      // FlutterResult — its Dart Future would simply never complete.
      result(FlutterError(code: "busy", message: "a pickFolder() call is already in progress", details: nil))
      return
    }
    guard let rootViewController = Self.rootViewController() else {
      result(FlutterError(code: "no_root_vc", message: "no root view controller to present the picker from", details: nil))
      return
    }

    let picker = UIDocumentPickerViewController(forOpeningContentTypes: [UTType.folder])
    picker.allowsMultipleSelection = false
    picker.delegate = self

    pendingPickCompletion = { [weak self] url in
      guard let self else { return }
      guard let url else {
        result(nil) // user cancelled
        return
      }
      guard url.startAccessingSecurityScopedResource() else {
        result(FlutterError(code: "scope_failed", message: "startAccessingSecurityScopedResource failed", details: nil))
        return
      }
      do {
        let bookmark = try url.bookmarkData()
        self.activateScope(url: url)
        result(["id": bookmark.base64EncodedString(), "displayName": url.lastPathComponent])
      } catch {
        url.stopAccessingSecurityScopedResource()
        result(FlutterError(code: "bookmark_failed", message: "\(error)", details: nil))
      }
    }
    rootViewController.present(picker, animated: true)
  }

  private static func rootViewController() -> UIViewController? {
    let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
    let activeScene = scenes.first { $0.activationState == .foregroundActive } ?? scenes.first
    if let windowScene = activeScene, let window = windowScene.windows.first(where: { $0.isKeyWindow }) ?? windowScene.windows.first {
      return window.rootViewController
    }
    return UIApplication.shared.delegate?.window??.rootViewController
  }

  // MARK: - restore

  private func restore(id: String) -> Bool {
    guard let bookmark = Data(base64Encoded: id) else { return false }
    var isStale = false
    guard
      let url = try? URL(resolvingBookmarkData: bookmark, options: [], relativeTo: nil, bookmarkDataIsStale: &isStale)
    else {
      return false
    }
    // Stale bookmark: the Dart side drops the persisted grant and
    // re-prompts via pickFolder() rather than trying to use a URL that may
    // no longer point at the right place.
    if isStale { return false }
    guard url.startAccessingSecurityScopedResource() else { return false }
    activateScope(url: url)
    return true
  }

  private func activateScope(url: URL) {
    activeURL?.stopAccessingSecurityScopedResource()
    activeURL = url
  }

  // MARK: - list

  private func list(id: String) -> [String] {
    guard let root = activeURL else { return [] }
    var out: [String] = []
    walk(dir: root, relPath: "", depth: Self.maxDepth, into: &out)
    return out
  }

  private func walk(dir: URL, relPath: String, depth: Int, into out: inout [String]) {
    guard depth > 0 else { return }
    guard
      let entries = try? FileManager.default.contentsOfDirectory(
        at: dir, includingPropertiesForKeys: [.isDirectoryKey], options: [.skipsHiddenFiles])
    else { return }

    for entry in entries {
      let name = entry.lastPathComponent
      if name.hasPrefix(".") { continue } // dotfiles/dot-dirs skipped, desktop rule
      let childRelPath = relPath.isEmpty ? name : "\(relPath)/\(name)"
      let isDir = (try? entry.resourceValues(forKeys: [.isDirectoryKey]))?.isDirectory ?? false
      if isDir {
        walk(dir: entry, relPath: childRelPath, depth: depth - 1, into: &out)
      } else if name.lowercased().hasSuffix(".md") || name.lowercased().hasSuffix(".markdown") {
        out.append(childRelPath)
      }
    }
  }

  // MARK: - read

  private func read(id: String, relPath: String) -> FlutterStandardTypedData? {
    guard let root = activeURL else { return nil }
    let target = root.appendingPathComponent(relPath)
    // Belt-and-suspenders: the Dart side (VaultPath.resolve) already
    // rejects any relPath that would walk above the vault root before it
    // ever reaches here, but refuse to serve anything outside the picked
    // root regardless of how the path arrived.
    //
    // A plain hasPrefix(root.path) is the classic sibling-directory flaw:
    // "/Users/x/VaultRoot-evil" has "/Users/x/VaultRoot" as a raw string
    // prefix without being inside it. Comparing against the root path with
    // a trailing separator closes that, while still accepting the target
    // legitimately BEING the root itself (relPath == "").
    let targetPath = target.standardizedFileURL.path
    let rootPath = root.standardizedFileURL.path
    let rootPathWithSeparator = rootPath.hasSuffix("/") ? rootPath : rootPath + "/"
    guard targetPath == rootPath || targetPath.hasPrefix(rootPathWithSeparator) else { return nil }
    guard let data = try? Data(contentsOf: target) else { return nil }
    return FlutterStandardTypedData(bytes: data)
  }
}

extension VaultChannel: UIDocumentPickerDelegate {
  func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
    pendingPickCompletion?(urls.first)
    pendingPickCompletion = nil
  }

  func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
    pendingPickCompletion?(nil)
    pendingPickCompletion = nil
  }
}
